server/src/main/java/keywhiz/service/daos/SecretContentDAO.java

```diff
@@ -130,6 +133,30 @@ public long createSecretContent(long secretId, String encryptedContent,
     return Optional.ofNullable(r).map(secretContentMapper::map);
   }
 
+  public Optional<ImmutableList<SecretContent>> getSecretVersionsBySecretId(long id, int newestIdx,
+      int oldestIdx) {
+    // Calculate how many rows to retrieve
+    if (oldestIdx == 0) {
+      // Retrieve all rows from newestIdx to the end
+      oldestIdx =
+          Integer.MAX_VALUE - 1; // -1 to compensate for + 1 in calculating range to retrieve
+    }
+
+    Result<SecretsContentRecord> r = dslContext.selectFrom(SECRETS_CONTENT)
+        .where(SECRETS_CONTENT.SECRETID.eq(id))
+        .orderBy(SECRETS_CONTENT.CREATEDAT.desc())
+        .limit(newestIdx, oldestIdx - newestIdx + 1)
```

Comment:
Given that mysql handles limits on query results as an offset + limit, why not implement your arguments (oldestIdx and newestIdx) in that way? Then the logic on 139-143 is unnecessary, you don't have to worry about negative values if oldest is less than newest, and (to me at least) the arguments to the API call will be more intuitive.